querqy-core/src/main/java/querqy/trie/RuleExtractorUtils.java

```diff
@@ -0,0 +1,38 @@
+package querqy.trie;
+
+import querqy.trie.model.ExactMatch;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+
+public class RuleExtractorUtils {
+
+    private RuleExtractorUtils(){}
+
+    private static final Comparator<ExactMatch<?>> COMPARE_BY_SIZE_DESC = (o1, o2) -> o2.termSize - o1.termSize;
+    private static final Comparator<ExactMatch<?>> COMPARE_BY_LOOKUP_START_ASC = Comparator.comparingInt(o -> o.lookupStart);
+
+    public static final Comparator<State<?>> COMPARE_STATE_BY_INDEX_DESC = Comparator.comparingInt(o -> o.index);
+
+    public static <T> List<ExactMatch<T>> removeSubsetsAndSmallerOverlaps(List<ExactMatch<T>> exactMatches) {
+        final List<ExactMatch<T>> exactMatchesFiltered = new ArrayList<>();
+
+        exactMatches.stream()
+                .sorted(COMPARE_BY_SIZE_DESC.thenComparing(COMPARE_BY_LOOKUP_START_ASC))
```

Comment:
What if the two terms have the same size and start at the same position? I'm not sure if this can happen in this context at all, but maybe we need to look at a tertiary criterion? The documentation of `sorted` says it's stable for sorted streams, which I don't think we can guarantee.